File: native/ios/NativeModules/NativeVRMSpringBones.swift
import RealityKit
import simd
import Foundation

internal struct ParsedSpringBoneCollider {
    let nodeName: String
    let offset: SIMD3<Float>
    let radius: Float
    let tail: SIMD3<Float>?
}

internal struct ParsedSpringBoneJoint {
    let nodeName: String
    let hitRadius: Float
    let stiffness: Float
    let gravityPower: Float
    let gravityDir: SIMD3<Float>
    let dragForce: Float
}

internal struct ParsedSpringBone {
    let name: String
    let centerNodeName: String?
    let joints: [ParsedSpringBoneJoint]
    let colliderIndices: [Int]
}

internal struct ParsedSpringBoneData {
    let colliders: [ParsedSpringBoneCollider]
    let springs: [ParsedSpringBone]
}

@available(iOS 18.0, *)
internal class NativeVRMSpringBones {
    class RuntimeCollider {
        let entity: Entity
        let offset: SIMD3<Float>
        let radius: Float
        let tail: SIMD3<Float>?
        var worldMatrix: float4x4 = matrix_identity_float4x4

        init(entity: Entity, parsed: ParsedSpringBoneCollider) {
            self.entity = entity
            self.offset = parsed.offset
            self.radius = parsed.radius
            self.tail = parsed.tail
        }
    }

    class RuntimeJoint {
        let entity: Entity
        let hitRadius: Float
        let stiffness: Float
        let gravityPower: Float
        let gravityDir: SIMD3<Float>
        let dragForce: Float
        let centerEntity: Entity?
        let colliders: [RuntimeCollider]

        var parentEntity: Entity?
        var childEntity: Entity?
        var initialTransform: float4x4 = matrix_identity_float4x4
        var localRestRot: simd_quatf = simd_quatf(ix: 0, iy: 0, iz: 0, r: 1)
        
        var boneAxis: SIMD3<Float> = SIMD3<Float>(0, -1, 0)
        var boneLength: Float = 0.05
        
        // Verlet state in WORLD space
        var currentTail: SIMD3<Float> = .zero
        var prevTail: SIMD3<Float> = .zero
        
        // Optimization: track motion to skip physics when idle
        var lastParentWorldPos: SIMD3<Float> = .zero
        var lastParentWorldRot: simd_quatf = simd_quatf(ix: 0, iy: 0, iz: 0, r: 1)
        var isSettled = false
        var initialized = false

        init(entity: Entity, parsed: ParsedSpringBoneJoint, center: Entity?, colliders: [RuntimeCollider]) {
            self.entity = entity
            self.hitRadius = parsed.hitRadius
            self.stiffness = parsed.stiffness
            self.gravityPower = parsed.gravityPower
            self.gravityDir = parsed.gravityDir
            self.dragForce = parsed.dragForce
            self.centerEntity = center
            self.colliders = colliders
        }
    }

    private var joints: [RuntimeJoint] = []
    private var colliders: [RuntimeCollider] = []
    private var accumulator: Float = 0
    private let fixedDeltaTime: Float = 1.0 / 60.0

    init(entitiesByName: [String: Entity], data: ParsedSpringBoneData) {
        var colliderMap: [Int: RuntimeCollider] = [:]
        for (index, parsedCol) in data.colliders.enumerated() {
            if let entity = entitiesByName[parsedCol.nodeName] {
                let rc = RuntimeCollider(entity: entity, parsed: parsedCol)
                colliders.append(rc)
                colliderMap[index] = rc
            }
        }

        let bodyKeywords = ["hair", "bust", "breast", "tail", "ear", "髪", "胸", "尾", "耳", "頭", "肢"]
        let outfitKeywords = ["skirt", "clothing", "cloth", "dress", "sleeve", "ribbon", "outfit", "accessory", "acc", "coat", "pant", "shirt", "tie", "cape", "apron", "hat", "jacket", "hoodie", "uniform", "waist", "belt", "pocket", "frill", "lace", "cap", "shoes", "boot", "sock", "onepiece", "necktie", "collar", "button", "mesh", "part", "prop", "extra", "wing", "feather", "bow", "ornament", "mantle", "shoulderpad", "armor", "decoration", "decor", "スカート", "服", "衣装", "袖", "リボン", "アクセ", "メッシュ", "羽", "翼", "飾り", "装飾"]

        for spring in data.springs {
            let centerEntity = spring.centerNodeName.flatMap { entitiesByName[$0] }
            let springColliders = spring.colliderIndices.compactMap { colliderMap[$0] }
            
            let lowerGroupName = spring.name.lowercased()
            let groupIsOutfit = outfitKeywords.contains { lowerGroupName.contains($0) }
            let groupIsBody = bodyKeywords.contains { lowerGroupName.contains($0) }

            for var parsedJoint in spring.joints {
                let lowerNodeName = parsedJoint.nodeName.lowercased()
                let nodeIsBody = bodyKeywords.contains { lowerNodeName.contains($0) }
                let nodeIsOutfit = outfitKeywords.contains { lowerNodeName.contains($0) }
                
                var finalIsOutfit = false
                if groupIsOutfit || nodeIsOutfit {
                    finalIsOutfit = true
                }
                // Body/Hair parts should take precedence to keep their gravity (hair points down)
                if groupIsBody || nodeIsBody {
                    finalIsOutfit = false
                }

                if finalIsOutfit {
                    // For outfits, we override the gravityPower to 0 to prevent sagging/drooping.
                    // We also increase dragForce to make them follow the body more rigidly without swinging.
                    let modifiedJoint = ParsedSpringBoneJoint(
                        nodeName: parsedJoint.nodeName,
                        hitRadius: parsedJoint.hitRadius,
                        stiffness: parsedJoint.stiffness * 0.5, // Reduced stiffness for smoother follow
                        gravityPower: 0,
                        gravityDir: parsedJoint.gravityDir,
                        dragForce: max(parsedJoint.dragForce, 0.8) // High drag for stable outfit follow
                    )
                    parsedJoint = modifiedJoint
                }

                if let entity = entitiesByName[parsedJoint.nodeName] {
                    let rj = RuntimeJoint(entity: entity, parsed: parsedJoint, center: centerEntity, colliders: springColliders)
                    joints.append(rj)
                }
            }
        }
    }

    func setup() {
        for joint in joints {
            joint.parentEntity = joint.entity.parent
            joint.childEntity = joint.entity.children.first

            joint.initialTransform = joint.entity.transform.matrix
            joint.localRestRot = joint.entity.transform.rotation

            if let child = joint.childEntity {
                let translation = child.transform.translation
                let len = length(translation)
                if len > 0.001 {
                    joint.boneAxis = normalize(translation)
                    joint.boneLength = len
                } else {
                    joint.boneAxis = SIMD3<Float>(0, -1, 0)
                    joint.boneLength = 0.05
                }
            } else {
                joint.boneAxis = SIMD3<Float>(0, -1, 0)
                joint.boneLength = 0.05
            }

            let worldPos = joint.entity.position(relativeTo: nil)
            let worldRot = joint.entity.orientation(relativeTo: nil)
            
            let localTail = joint.boneAxis * joint.boneLength
            let worldTailPos = worldPos + worldRot.act(localTail)

            joint.currentTail = worldTailPos
            joint.prevTail = worldTailPos
            joint.lastParentWorldPos = worldPos
            joint.lastParentWorldRot = worldRot
            joint.initialized = true
        }
    }

    func update(deltaTime: Float) {
        if deltaTime <= 0 || joints.isEmpty { return }

        // Update colliders
        for col in colliders {
            col.worldMatrix = col.entity.transformMatrix(relativeTo: nil)
        }

        accumulator += deltaTime
        // Limit accumulator to prevent "spiral of death" during lag spikes
        if accumulator > 0.1 { accumulator = 0.1 }

        // Use 2 sub-steps for better stability (120Hz effective physics)
        let subStepDt = fixedDeltaTime * 0.5
        while accumulator >= subStepDt {
            performPhysicsStep(dt: subStepDt)
            accumulator -= subStepDt
        }
    }

    private func performPhysicsStep(dt: Float) {
        for joint in joints {
            guard joint.initialized else { continue }

            let parentTransform = joint.parentEntity?.transformMatrix(relativeTo: nil) ?? matrix_identity_float4x4
            let localPos = joint.initialTransform.columns.3
            let currentParentWorldPos4 = parentTransform * localPos
            let nWorldPos = SIMD3<Float>(currentParentWorldPos4.x, currentParentWorldPos4.y, currentParentWorldPos4.z)
            
            let worldRot = (joint.parentEntity?.orientation(relativeTo: nil) ?? simd_quatf(ix: 0, iy: 0, iz: 0, r: 1)) * joint.localRestRot
            
            // Motion Check with slightly higher epsilon for stability
            let moveDist = length(nWorldPos - joint.lastParentWorldPos)
            let rotDist = length(worldRot.vector - joint.lastParentWorldRot.vector)
            
            if moveDist > 0.0002 || rotDist > 0.0002 {
                joint.isSettled = false
                joint.lastParentWorldPos = nWorldPos
                joint.lastParentWorldRot = worldRot
            }

            let localRestCombined = parentTransform * joint.initialTransform
            let restTailWorldVec4 = localRestCombined * SIMD4<Float>(joint.boneAxis * joint.boneLength, 0.0)
            let restTailWorld = nWorldPos + SIMD3<Float>(restTailWorldVec4.x, restTailWorldVec4.y, restTailWorldVec4.z)

            if joint.isSettled {
                joint.currentTail = restTailWorld
                joint.prevTail = restTailWorld
                continue
            }

            // 1. Verlet Integration
            let inertia = (joint.currentTail - joint.prevTail) * (1.0 - joint.dragForce)
            let gravity = joint.gravityDir * (joint.gravityPower * 9.81 * dt * dt)
            let stiffnessPull = (restTailWorld - joint.currentTail) * (joint.stiffness * dt)

            var nextTail = joint.currentTail + inertia + gravity + stiffnessPull

            // 2. Bone Length Constraint & Velocity Capping
            // Prevents "vibrating" when stiffness is too high or bones are compressed.
            let toNext = nextTail - nWorldPos
            let toNextLen = length(toNext)
            if toNextLen > 0.0001 {
                // Cap the maximum displacement from rest position per step (stability)
                let maxShift = joint.boneLength * 0.25
                let shift = length(nextTail - joint.currentTail)
                if shift > maxShift {
                    nextTail = joint.currentTail + normalize(nextTail - joint.currentTail) * maxShift
                }
                
                // Final length constraint
                nextTail = nWorldPos + normalize(nextTail - nWorldPos) * joint.boneLength
            }

            // 3. Collision
            for col in joint.colliders {
                nextTail = resolveCollision(tail: nextTail, nodePos: nWorldPos, boneLength: joint.boneLength, hitRadius: joint.hitRadius, col: col)
            }

            // 4. Settling Refinement
            let velocity = length(nextTail - joint.currentTail)
            let distFromRest = length(nextTail - restTailWorld)
            // Settling helps stop microscopic oscillations (vibrations)
            if velocity < 0.0005 && distFromRest < 0.001 {
                joint.isSettled = true
            }

            joint.prevTail = joint.currentTail
            joint.currentTail = nextTail

            // 5. Apply Rotation (Stable projection)
            let parentInv = parentTransform.inverse
            let localNextTailParent4 = parentInv * SIMD4<Float>(nextTail.x, nextTail.y, nextTail.z, 1.0)
            let localNextTailParent = SIMD3<Float>(localNextTailParent4.x, localNextTailParent4.y, localNextTailParent4.z)
            
            let localP = SIMD3<Float>(localPos.x, localPos.y, localPos.z)
            let diff = localNextTailParent - localP
            if length(diff) > 0.0001 {
                let toTargetLocal = normalize(diff)
                let boneAxisParentSpace = normalize(joint.localRestRot.act(joint.boneAxis))

                let rotAxis = cross(boneAxisParentSpace, toTargetLocal)
                let rotAxisLen = length(rotAxis)
                let dotP = max(-1.0, min(1.0, dot(boneAxisParentSpace, toTargetLocal)))
                let rotAngle = acos(dotP)

                var localDeltaRot = simd_quatf(ix: 0, iy: 0, iz: 0, r: 1)
                if rotAxisLen > 0.0001 && rotAngle > 0.0001 {
                    localDeltaRot = simd_quatf(angle: rotAngle, axis: normalize(rotAxis))
                }
                joint.entity.transform.rotation = localDeltaRot * joint.localRestRot
            }
        }
    }

    private func resolveCollision(tail: SIMD3<Float>, nodePos: SIMD3<Float>, boneLength: Float, hitRadius: Float, col: RuntimeCollider) -> SIMD3<Float> {
        var currentTail = tail
        let m = col.worldMatrix

        if let tailEnd = col.tail {
            let start4 = m * SIMD4<Float>(col.offset.x, col.offset.y, col.offset.z, 1.0)
            let end4 = m * SIMD4<Float>(tailEnd.x, tailEnd.y, tailEnd.z, 1.0)
            let start = SIMD3<Float>(start4.x, start4.y, start4.z)
            let end = SIMD3<Float>(end4.x, end4.y, end4.z)
            let r = col.radius + hitRadius

            let segment = end - start
            let segmentLenSq = dot(segment, segment)
            var t: Float = 0
            if segmentLenSq > 0.0001 {
                t = max(0, min(1, dot(currentTail - start, segment) / segmentLenSq))
            }
            let center = start + segment * t

            let distVec = currentTail - center
            let dist = length(distVec)
            if dist < r && dist > 0.0001 {
                let push = normalize(distVec) * r
                currentTail = center + push
                currentTail = nodePos + normalize(currentTail - nodePos) * boneLength
            }
        } else {
            let center4 = m * SIMD4<Float>(col.offset.x, col.offset.y, col.offset.z, 1.0)
            let center = SIMD3<Float>(center4.x, center4.y, center4.z)
            let r = col.radius + hitRadius
            let distVec = currentTail - center
            let dist = length(distVec)
            if dist < r && dist > 0.0001 {
                let push = normalize(distVec) * r
                currentTail = center + push
                currentTail = nodePos + normalize(currentTail - nodePos) * boneLength
            }
        }

        return currentTail
    }
}

File: native/android/vrm/NativeVrmSpringBones.kt
package com.dedicatus.VroidViewer.vrm

import android.util.Log
import com.google.android.filament.TransformManager
import dev.romainguy.kotlin.math.*
import kotlin.math.acos

internal fun extractTranslation(m: Mat4): Float3 {
    return Float3(m.w.x, m.w.y, m.w.z)
}

internal fun mulMat4Float4(m: Mat4, v: Float4): Float4 {
    val vx = v.x; val vy = v.y; val vz = v.z; val vw = v.w
    return Float4(
        m.x.x * vx + m.y.x * vy + m.z.x * vz + m.w.x * vw,
        m.x.y * vx + m.y.y * vy + m.z.y * vz + m.w.y * vw,
        m.x.z * vx + m.y.z * vy + m.z.z * vz + m.w.z * vw,
        m.x.w * vx + m.y.w * vy + m.z.w * vz + m.w.w * vw
    )
}

internal fun mulQuatFloat3(q: Quaternion, v: Float3): Float3 {
    val qx = q.x; val qy = q.y; val qz = q.z; val qw = q.w
    val vx = v.x; val vy = v.y; val vz = v.z
    val tx = 2f * (qy * vz - qz * vy)
    val ty = 2f * (qz * vx - qx * vz)
    val tz = 2f * (qx * vy - qy * vx)
    val cx = qy * tz - qz * ty
    val cy = qz * tx - qx * tz
    val cz = qx * ty - qy * tx
    return Float3(
        vx + qw * tx + cx,
        vy + qw * ty + cy,
        vz + qw * tz + cz
    )
}

internal sealed class SpringBoneColliderShape {
    data class Sphere(val offset: Float3, val radius: Float) : SpringBoneColliderShape()
    data class Capsule(val offset: Float3, val radius: Float, val tail: Float3) : SpringBoneColliderShape()
}

internal data class ParsedSpringBoneCollider(
    val nodeName: String,
    val shape: SpringBoneColliderShape
)

internal data class ParsedSpringBoneJoint(
    val nodeName: String,
    val hitRadius: Float,
    val stiffness: Float,
    val gravityPower: Float,
    val gravityDir: Float3,
    val dragForce: Float
)

internal data class ParsedSpringBone(
    val name: String,
    val centerNodeName: String?,
    val joints: List<ParsedSpringBoneJoint>,
    val colliderIndices: List<Int> // Indices pointing into the global ParsedSpringBoneCollider list
)

internal data class ParsedSpringBoneData(
    val colliders: List<ParsedSpringBoneCollider>,
    val springs: List<ParsedSpringBone>
)

internal class NativeVrmSpringBones(
    private val transformManager: TransformManager,
    private val boneEntities: Map<String, Int>,
    val data: ParsedSpringBoneData
) {
    private val joints = mutableListOf<RuntimeJoint>()
    private val colliders = mutableListOf<RuntimeCollider>()

    inner class RuntimeCollider(
        val entity: Int,
        val shape: SpringBoneColliderShape
    ) {
        var worldMatrix: Mat4 = Mat4()
    }

    inner class RuntimeJoint(
        val entity: Int,
        val hitRadius: Float,
        val stiffness: Float,
        val gravityPower: Float,
        val gravityDir: Float3,
        val dragForce: Float,
        val colliders: List<RuntimeCollider>,
        val centerEntity: Int?
    ) {
        var parentInst: Int = 0
        var childInst: Int = 0 
        
        // Verlet state in WORLD space
        var currentTail = Float3()
        var prevTail = Float3()

        // Rest pose state
        var boneAxis = Float3(0f, 1f, 0f)
        var boneLength = 0f
        var localRestRot = Quaternion()
        var initialTransform = Mat4()

        // Optimization: track parent motion to skip physics when idle
        var lastParentWorldPos = Float3()
        var isSettled = false
        var initialized = false
    }

    init {
        val colliderMap = mutableMapOf<Int, RuntimeCollider>()
        data.colliders.forEachIndexed { index, parsedCol ->
            val entity = boneEntities[parsedCol.nodeName]
            if (entity != null) {
                val rc = RuntimeCollider(entity, parsedCol.shape)
                colliders.add(rc)
                colliderMap[index] = rc
            }
        }

        val bodyKeywords = listOf("hair", "bust", "breast", "tail", "ear", "髪", "胸", "尾", "耳")
        val outfitKeywords = listOf("skirt", "clothing", "cloth", "dress", "sleeve", "ribbon", "outfit", "accessory", "acc", "coat", "pant", "shirt", "tie", "cape", "apron", "hat", "jacket", "hoodie", "uniform", "waist", "belt", "pocket", "frill", "lace", "cap", "shoes", "boot", "sock", "onepiece", "necktie", "collar", "button", "スカート", "服", "衣装", "袖", "リボン", "アクセ")

        for (spring in data.springs) {
            val centerEntity = spring.centerNodeName?.let { boneEntities[it] }
            val springColliders = spring.colliderIndices.mapNotNull { colliderMap[it] }

            val lowerGroupName = spring.name.lowercase()
            val groupIsOutfit = outfitKeywords.any { lowerGroupName.contains(it) }
            val groupIsBody = bodyKeywords.any { lowerGroupName.contains(it) }

            for (parsedJoint in spring.joints) {
                val entity = boneEntities[parsedJoint.nodeName] ?: continue
                
                val lowerNodeName = parsedJoint.nodeName.lowercase()
                val nodeIsBody = bodyKeywords.any { lowerNodeName.contains(it) }
                val nodeIsOutfit = outfitKeywords.any { lowerNodeName.contains(it) }

                var finalIsOutfit = false
                if (groupIsOutfit || nodeIsOutfit) {
                    finalIsOutfit = true
                }
                // Body/Hair parts should take precedence to keep their gravity (hair points down)
                if (groupIsBody || nodeIsBody) {
                    finalIsOutfit = false
                }

                // The gravity should affect the body part (hair/bust) but not the outfit.
                // We override the gravityPower to 0 for identified outfits to prevent sagging/drooping.
                val gravityPower = if (finalIsOutfit) 0f else parsedJoint.gravityPower

                val rj = RuntimeJoint(
                    entity = entity,
                    hitRadius = parsedJoint.hitRadius,
                    stiffness = parsedJoint.stiffness,
                    gravityPower = gravityPower,
                    gravityDir = parsedJoint.gravityDir,
                    dragForce = parsedJoint.dragForce,
                    colliders = springColliders,
                    centerEntity = centerEntity
                )
                joints.add(rj)
            }
        }
    }

    private fun safeRotation(m: Mat4): Quaternion {
        val trace = m.x.x + m.y.y + m.z.z
        if (trace > 0.0f) {
            val s = 0.5f / kotlin.math.sqrt(trace + 1.0f)
            return Quaternion(
                (m.y.z - m.z.y) * s,
                (m.z.x - m.x.z) * s,
                (m.x.y - m.y.x) * s,
                0.25f / s
            )
        } else {
            if (m.x.x > m.y.y && m.x.x > m.z.z) {
                val s = 2.0f * kotlin.math.sqrt(1.0f + m.x.x - m.y.y - m.z.z)
                return Quaternion(
                    0.25f * s,
                    (m.x.y + m.y.x) / s,
                    (m.x.z + m.z.x) / s,
                    (m.y.z - m.z.y) / s
                )
            } else if (m.y.y > m.z.z) {
                val s = 2.0f * kotlin.math.sqrt(1.0f + m.y.y - m.x.x - m.z.z)
                return Quaternion(
                    (m.x.y + m.y.x) / s,
                    0.25f * s,
                    (m.y.z + m.z.y) / s,
                    (m.z.x - m.x.z) / s
                )
            } else {
                val s = 2.0f * kotlin.math.sqrt(1.0f + m.z.z - m.x.x - m.y.y)
                return Quaternion(
                    (m.x.z + m.z.x) / s,
                    (m.y.z + m.z.y) / s,
                    0.25f * s,
                    (m.x.y - m.y.x) / s
                )
            }
        }
    }

    fun setup() {
        val tm = transformManager
        Log.i("NativeVrmSpringBones", "Setting up \${joints.size} SpringBone joints...")
        
        for (joint in joints) {
            val inst = tm.getInstance(joint.entity)
            if (inst == 0) continue

            val parentInst = tm.getParent(inst)
            joint.parentInst = parentInst

            val childCount = tm.getChildCount(inst)
            if (childCount > 0) {
                val childrenHolder = IntArray(childCount)
                tm.getChildren(inst, childrenHolder)
                joint.childInst = childrenHolder[0]
            }

            val matArr = FloatArray(16)
            tm.getTransform(inst, matArr)
            joint.initialTransform = Mat4.of(*matArr)
            joint.localRestRot = safeRotation(joint.initialTransform)

            if (joint.childInst != 0) {
                val childInst = joint.childInst
                val childTransformArr = FloatArray(16)
                tm.getTransform(childInst, childTransformArr)
                val tx = extractTranslation(Mat4.of(*childTransformArr))
                val len = length(tx)
                if (len > 0.001f) {
                    joint.boneAxis = normalize(tx)
                    joint.boneLength = len
                } else {
                    joint.boneAxis = Float3(0f, -1f, 0f) // Standard VRM hair points down
                    joint.boneLength = 0.05f
                }
            } else {
                joint.boneAxis = Float3(0f, -1f, 0f)
                joint.boneLength = 0.05f
            }

            val worldTransformArr = FloatArray(16)
            tm.getWorldTransform(inst, worldTransformArr)
            val worldTransform = Mat4.of(*worldTransformArr)
            
            val tailOffset = Float3(joint.boneAxis.x * joint.boneLength, joint.boneAxis.y * joint.boneLength, joint.boneAxis.z * joint.boneLength)
            val wtp4 = mulMat4Float4(worldTransform, Float4(tailOffset.x, tailOffset.y, tailOffset.z, 1f))
            val worldTailPos = Float3(wtp4.x, wtp4.y, wtp4.z)

            joint.currentTail = worldTailPos
            joint.prevTail = worldTailPos
            joint.lastParentWorldPos = extractTranslation(worldTransform)
            joint.initialized = true
        }
    }

    private var accumulator = 0f
    private val fixedDeltaTime = 1f / 60f

    fun update(deltaTime: Float) {
        if (deltaTime <= 0f || joints.isEmpty()) return
        val tm = transformManager

        // Sub-stepping for stability
        accumulator += deltaTime
        if (accumulator > 0.5f) accumulator = 0.5f // Cap to avoid "clock spiral of death"
        
        // Pre-fetch all collider matrices once per update
        for (col in colliders) {
            val inst = tm.getInstance(col.entity)
            if (inst != 0) {
                val floats = FloatArray(16)
                tm.getWorldTransform(inst, floats)
                col.worldMatrix = Mat4.of(*floats)
            }
        }

        while (accumulator >= fixedDeltaTime) {
            performPhysicsStep(fixedDeltaTime)
            accumulator -= fixedDeltaTime
        }
    }

    private fun performPhysicsStep(dt: Float) {
        val tm = transformManager
        
        for (joint in joints) {
            if (!joint.initialized) continue
            val inst = tm.getInstance(joint.entity)
            if (inst == 0) continue
            val parentInst = joint.parentInst

            var parentTransform = Mat4()
            if (parentInst != 0) {
                val pFloats = FloatArray(16)
                tm.getWorldTransform(parentInst, pFloats)
                parentTransform = Mat4.of(*pFloats)
            }

            val itTrans = extractTranslation(joint.initialTransform)
            val cpwp4 = mulMat4Float4(parentTransform, Float4(itTrans.x, itTrans.y, itTrans.z, 1f))
            val currentParentWorldPos = Float3(cpwp4.x, cpwp4.y, cpwp4.z)
            
            // Motion Check
            val diffP = Float3(currentParentWorldPos.x - joint.lastParentWorldPos.x, currentParentWorldPos.y - joint.lastParentWorldPos.y, currentParentWorldPos.z - joint.lastParentWorldPos.z)
            val moveDist = length(diffP)
            
            if (moveDist > 0.0001f) {
                joint.isSettled = false
                joint.lastParentWorldPos = currentParentWorldPos
            }

            // Calculate rest pose in world space
            val localRestCombined = parentTransform * joint.initialTransform
            val baScale = Float3(joint.boneAxis.x * joint.boneLength, joint.boneAxis.y * joint.boneLength, joint.boneAxis.z * joint.boneLength)
            val rtw4 = mulMat4Float4(localRestCombined, Float4(baScale.x, baScale.y, baScale.z, 1f))
            val restTailWorld = Float3(rtw4.x, rtw4.y, rtw4.z)

            if (joint.isSettled) {
                joint.currentTail = restTailWorld
                joint.prevTail = restTailWorld
                continue 
            }

            // 1. Verlet Integration
            val dragScale = 1.0f - joint.dragForce
            val diffPrev = Float3(joint.currentTail.x - joint.prevTail.x, joint.currentTail.y - joint.prevTail.y, joint.currentTail.z - joint.prevTail.z)
            val inertia = Float3(diffPrev.x * dragScale, diffPrev.y * dragScale, diffPrev.z * dragScale)

            val gScaled = joint.gravityPower * 9.81f * dt * dt
            val gravityVector = Float3(joint.gravityDir.x * gScaled, joint.gravityDir.y * gScaled, joint.gravityDir.z * gScaled)

            val stiffScale = joint.stiffness * dt
            val diffStiff = Float3(restTailWorld.x - joint.currentTail.x, restTailWorld.y - joint.currentTail.y, restTailWorld.z - joint.currentTail.z)
            val stiffness = Float3(diffStiff.x * stiffScale, diffStiff.y * stiffScale, diffStiff.z * stiffScale)

            var nextTail = Float3(
                joint.currentTail.x + inertia.x + gravityVector.x + stiffness.x,
                joint.currentTail.y + inertia.y + gravityVector.y + stiffness.y,
                joint.currentTail.z + inertia.z + gravityVector.z + stiffness.z
            )

            // 2. Bone Length Constraint
            val toNext = Float3(nextTail.x - currentParentWorldPos.x, nextTail.y - currentParentWorldPos.y, nextTail.z - currentParentWorldPos.z)
            val toNextLen = length(toNext)
            if (toNextLen > 0.0001f) {
                val dirX = toNext.x / toNextLen
                val dirY = toNext.y / toNextLen
                val dirZ = toNext.z / toNextLen
                nextTail = Float3(
                    currentParentWorldPos.x + dirX * joint.boneLength,
                    currentParentWorldPos.y + dirY * joint.boneLength,
                    currentParentWorldPos.z + dirZ * joint.boneLength
                )
            }

            // 3. Collision Resolution
            for (col in joint.colliders) {
                nextTail = resolveCollision(nextTail, currentParentWorldPos, joint.boneLength, joint.hitRadius, col)
            }

            // 4. Settling Logic
            val velVec = Float3(nextTail.x - joint.currentTail.x, nextTail.y - joint.currentTail.y, nextTail.z - joint.currentTail.z)
            val velocity = length(velVec)
            val distRestVec = Float3(nextTail.x - restTailWorld.x, nextTail.y - restTailWorld.y, nextTail.z - restTailWorld.z)
            val distFromRest = length(distRestVec)
            if (velocity < 0.0001f && distFromRest < 0.001f) {
                joint.isSettled = true
            }

            joint.prevTail = joint.currentTail
            joint.currentTail = nextTail

            // 5. Apply Rotation back to Filament
            val lntp4 = mulMat4Float4(inverse(parentTransform), Float4(nextTail.x, nextTail.y, nextTail.z, 1f))
            val lntp3 = Float3(lntp4.x, lntp4.y, lntp4.z)
            val tj = extractTranslation(joint.initialTransform)
            val diffTLocal = Float3(lntp3.x - tj.x, lntp3.y - tj.y, lntp3.z - tj.z)
            val toTargetLocal = normalize(diffTLocal)
            val boneAxisParentSpace = normalize(mulQuatFloat3(joint.localRestRot, joint.boneAxis))
            
            val rotAxis = normalize(cross(boneAxisParentSpace, toTargetLocal))
            val dotP = clamp(dot(boneAxisParentSpace, toTargetLocal), -1f, 1f)
            val rotAngle = acos(dotP)
            
            var localDeltaRot = Quaternion()
            if (length(rotAxis) > 0.001f && rotAngle > 0.001f) {
                val halfAngle = rotAngle * 0.5f
                val s = kotlin.math.sin(halfAngle)
                localDeltaRot = Quaternion(rotAxis.x * s, rotAxis.y * s, rotAxis.z * s, kotlin.math.cos(halfAngle))
            }

            val finalRot = localDeltaRot * joint.localRestRot
            val matR = finalRot.toMatrix()
            val matT = translation(tj)
            val newTx = matT * matR // scale is 1f
            tm.setTransform(inst, newTx.toFloatArray())
        }
    }

    private fun resolveCollision(
        tail: Float3, 
        nodePos: Float3, 
        boneLength: Float, 
        hitRadius: Float,
        col: RuntimeCollider
    ): Float3 {
        var currentTail = tail
        val m = col.worldMatrix

        when (val s = col.shape) {
            is SpringBoneColliderShape.Sphere -> {
                val c4 = mulMat4Float4(m, Float4(s.offset.x, s.offset.y, s.offset.z, 1f))
                val center = Float3(c4.x, c4.y, c4.z)
                val r = s.radius + hitRadius
                val distVec = Float3(currentTail.x - center.x, currentTail.y - center.y, currentTail.z - center.z)
                val dist = length(distVec)
                if (dist < r && dist > 0.0001f) {
                    val pushDir = normalize(distVec)
                    val push = Float3(pushDir.x * r, pushDir.y * r, pushDir.z * r)
                    currentTail = Float3(center.x + push.x, center.y + push.y, center.z + push.z)
                    // Re-enforce bone length after pushing out
                    val diff = Float3(currentTail.x - nodePos.x, currentTail.y - nodePos.y, currentTail.z - nodePos.z)
                    val diffDir = normalize(diff)
                    currentTail = Float3(nodePos.x + diffDir.x * boneLength, nodePos.y + diffDir.y * boneLength, nodePos.z + diffDir.z * boneLength)
                }
            }
            is SpringBoneColliderShape.Capsule -> {
                val st4 = mulMat4Float4(m, Float4(s.offset.x, s.offset.y, s.offset.z, 1f))
                val start = Float3(st4.x, st4.y, st4.z)
                val end4 = mulMat4Float4(m, Float4(s.tail.x, s.tail.y, s.tail.z, 1f))
                val end = Float3(end4.x, end4.y, end4.z)
                val r = s.radius + hitRadius

                // Line segment projection
                val segment = Float3(end.x - start.x, end.y - start.y, end.z - start.z)
                val segmentLenSq = dot(segment, segment)
                var t = 0f
                if (segmentLenSq > 0.0001f) {
                    val diffC = Float3(currentTail.x - start.x, currentTail.y - start.y, currentTail.z - start.z)
                    t = clamp(dot(diffC, segment) / segmentLenSq, 0f, 1f)
                }
                val center = Float3(start.x + segment.x * t, start.y + segment.y * t, start.z + segment.z * t)

                val distVec = Float3(currentTail.x - center.x, currentTail.y - center.y, currentTail.z - center.z)
                val dist = length(distVec)
                if (dist < r && dist > 0.0001f) {
                    val pushDir = normalize(distVec)
                    val push = Float3(pushDir.x * r, pushDir.y * r, pushDir.z * r)
                    currentTail = Float3(center.x + push.x, center.y + push.y, center.z + push.z)
                    val diff = Float3(currentTail.x - nodePos.x, currentTail.y - nodePos.y, currentTail.z - nodePos.z)
                    val diffDir = normalize(diff)
                    currentTail = Float3(nodePos.x + diffDir.x * boneLength, nodePos.y + diffDir.y * boneLength, nodePos.z + diffDir.z * boneLength)
                }
            }
        }

        return currentTail
    }
}
